doc/developer/design/20230519_statement_logging.md

```diff
@@ -0,0 +1,165 @@
+# Statement Logging (Chapter One)
+
+<!-- - Associated: (Insert list of associated epics, issues, or PRs) -->
+
+<!-- <\!-- -->
+<!-- Note: Feel free to add or remove sections as needed. However, most design -->
+<!-- docs should at least keep the suggested sections. -->
+<!-- -\-> -->
+
+## Context
+
+There exists no -- or very limited -- visiblity into the body of
+statements that users are running against Materialize, either for us
+or for the users themselves. This creates several problems. Some
+examples, in no particular order:
+
+* We can't allocate engineering resources effectively. For example, we
+  don't know which query patterns are particularly common and thus
+  that we should put particular effort into optimizing.
+* Administrators in large organizations have no way to see what
+  queries people (or machines) in their organization are commonly
+  issuing
+* If a user's cluster is crashing, support (or the users themselves)
+  are not able to attempt to debug by checking whether a particularly
+  expensive or dangerous query was run recently.
+
+Note that two meaningfully distinct features arise from these. The
+first is a user-visible statement log; that feature is the subject of
+this design doc ("Chapter One"). An upcoming design doc ("Chapter
+Two") will discuss the second feature: logging statements in _our_
+infrastructure, for our own benefit. It is thought that implementing
+Chapter One is a prerequisite for implementing Chapter Two.
+
+## Goals
+
+* Administrators should be able to access a table that records all
+  statements issued in the recent past (e.g., last 30 days), along
+  with other relevvant information (e.g., timestamp information, time
+  to return, current cluster, current user, plan information, etc.)
+* (Stretch goal) unprivileged users should be able to access a view
+  that filters `mz_statement_history` to only surface queries that
+  that user has 
+
+## Non-Goals
+
+* Chapter Two (getting the queries into our systems for our own
+  analytics) is explicitly out of scope of _this_ design, but is
+  necessary to unlock its full value, and so should be done as a
+  follow-up ASAP.
+* 100% durability of queries is permanently out of scope. We have to
+  accept the possibility of losing data under adverse conditions. The
+  reason for this design choice is that otherwise we would have to
+  persist the text of every statement (in either Cockroach or S3),
+  which we consider unacceptable for `SELECT` queries.
+
+## Overview
+
+The coordinator maintains a buffer of pending events related to
+statement execution. Whenever a prepared statement is created, a
+corresponding event is added to the buffer. Whenever a statement is
+executed, a system var called `STATEMENT_LOG_SAMPLING_RATE` is
```

Comment:
PostgreSQL calls their parameter `log_statement_sample_rate`. I think we should use the "sample rate" phrasing rather than "sampling rate" for consistency. Maybe `statement_history_sample_rate`, for consistency with the table they'll be recorded in, and also to distance ourselves from the PostgreSQL feature, which writes the statements to the log file rather than to a table?